langwatch/src/server/tracer/opentelemetry.ts

```diff
@@ -174,12 +179,31 @@ const addOpenTelemetrySpanAsSpan = (
   // First token at
   let first_token_at: Span["timestamps"]["first_token_at"] = null;
   for (const event of otelSpan?.events ?? []) {
-    if (
-      event?.name === "First Token Stream Event" ||
-      event?.name === "llm.content.completion.chunk"
-    ) {
-      first_token_at = parseTimestamp(event?.timeUnixNano);
-      break;
+    if (!event) continue;
+
+    switch (event.name) {
+      case "First Token Stream Event":
+      case "llm.content.completion.chunk":
+        first_token_at = parseTimestamp(event?.timeUnixNano);
+        break;
+      case "langwatch.evaluation.custom":
+        const jsonPayload = event.attributes?.find((attr) => attr?.key === "json_encoded_event")?.value?.stringValue;
+        if (!jsonPayload) {
+          logger.warn({ event }, "event for `langwatch.evaluation.custom` has no json_encoded_event");
+          break;
+        }
+
+        try {
+          const evaluation = JSON.parse(jsonPayload) as Evaluation;
```

Comment:
to not risk saving invalid data, we shouldn't just trust the incoming data and force type conversion with `as`, instead we should actually do `rESTEvaluationSchema.parse`

(and if there is an error use `fromZodError` before logging otherwise it can be quite unreadable)

Reply:
it might actually fail on elasticsearch schema level if we don't do this

Reply:
Ahh nice, I was going to ask you about this. Will sort tomorrow!